aws/resource_aws_sagemaker_endpoint.go

```diff
@@ -0,0 +1,224 @@
+			log.Printf("[INFO] unable to find the sagemaker endpoint resource and therefore it is removed from the state: %s", d.Id())
+			d.SetId("")
+			return nil
+		}
+		return err
+	}
+	if endpointRaw == nil {
+		d.SetId("")
+		return nil
+	}
+
+	endpoint := endpointRaw.(*sagemaker.DescribeEndpointOutput)
+
+	if err := d.Set("name", endpoint.EndpointName); err != nil {
+		return err
+	}
+	if err := d.Set("endpoint_config_name", endpoint.EndpointConfigName); err != nil {
+		return err
+	}
+
+	if err := d.Set("arn", endpoint.EndpointArn); err != nil {
+		return err
+	}
+
+	tagsOutput, err := conn.ListTags(&sagemaker.ListTagsInput{
+		ResourceArn: endpoint.EndpointArn,
+	})
+	if err := d.Set("tags", tagsToMapSagemaker(tagsOutput.Tags)); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func resourceAwsSagemakerEndpointUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).sagemakerconn
+
+	d.Partial(true)
+
+	if err := setSagemakerTags(conn, d); err != nil {
+		return err
+	}
+	d.SetPartial("tags")
+
+	if d.HasChange("endpoint_config_name") && !d.IsNewResource() {
+		modifyOpts := &sagemaker.UpdateEndpointInput{
+			EndpointName:       aws.String(d.Id()),
+			EndpointConfigName: aws.String(d.Get("endpoint_config_name").(string)),
+		}
+		log.Printf(
+			"[INFO] Modifying endpoint_config_name attribute for %s: %#v",
+			d.Id(), modifyOpts)
+		if _, err := conn.UpdateEndpoint(modifyOpts); err != nil {
+			return err
+		}
+		d.SetPartial("endpoint_config_name")
+
+		log.Printf("[DEBUG] Waiting for SageMaker endpoint (%s) to be updated", d.Id())
+		stateConf := &resource.StateChangeConf{
+			Pending: []string{"Updating"},
+			Target:  []string{"InService"},
+			Refresh: SagemakerEndpointStateRefreshFunc(conn, d.Id()),
+			Timeout: 15 * time.Minute,
+		}
+		if _, err := stateConf.WaitForState(); err != nil {
+			return fmt.Errorf(
+				"error updating SageMaker endpoint (%s): %s",
+				d.Id(), err)
+		}
+	}
+
+	d.Partial(false)
+
+	return resourceAwsSagemakerEndpointRead(d, meta)
+}
+
+func resourceAwsSagemakerEndpointDelete(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).sagemakerconn
+
+	deleteEndpointOpts := &sagemaker.DeleteEndpointInput{
+		EndpointName: aws.String(d.Id()),
+	}
+	log.Printf("[INFO] Deleting Sagemaker endpoint: %s", d.Id())
+
+	return resource.Retry(5*time.Minute, func() *resource.RetryError {
+		_, err := conn.DeleteEndpoint(deleteEndpointOpts)
+		if err == nil {
+			return nil
+		}
+
+		sagemakerErr, ok := err.(awserr.Error)
+		if !ok {
+			return resource.NonRetryableError(err)
+		}
+
+		if sagemakerErr.Code() == "ResourceNotFound" {
+			return resource.RetryableError(err)
+		}
+
+		return resource.NonRetryableError(fmt.Errorf("Error deleting Sagemaker endpoint: %s", err))
+	})
+}
+
+func SagemakerEndpointStateRefreshFunc(conn *sagemaker.SageMaker, name string) resource.StateRefreshFunc {
```

Comment:
Nit: If we keep this function (given potential removal with the AWS Go SDK waiter functions), we should prefer to not export our Go functions as this codebase is not intended for library usage.

```suggestion
func sagemakerEndpointStateRefreshFunc(conn *sagemaker.SageMaker, name string) resource.StateRefreshFunc {
```